### ISSUE TITLE
Direction of current seems wrong

### ISSUE THREAD

**Issue by svejgaard** [bfieldtools/bfieldtools#35]
Hi guys,
Thank you very much for sharing the library!
I am having troubles regarding the direction of the current in the coil loops (and the direction of the resulting magnetic field).
I am demonstrating the issue using your official examples as much as possible. The setup of the design space follows the MAMBA coil example, but I only create one single target point, with its magnetic field pointing in the positive X direction.
Then I calculate and plot the magnetic field using the procedure from Biplanar coil design example and Magnetically shielded coil. But the result comes out either with the magnetic field in the opposite direction of the defined target field, or with wrong direction of current compared to calculated magnetic field.

```python
import numpy as np
from mayavi import mlab
%gui qt # for mayavi to work in Spyder
import trimesh
from bfieldtools.mesh_conductor import MeshConductor
from bfieldtools.coil_optimize import optimize_streamfunctions
from bfieldtools.contour import scalar_contour
from bfieldtools.viz import plot_3d_current_loops
from bfieldtools.utils import combine_meshes, load_example_mesh

# Load simple plane mesh that is centered on the origin
planemesh = load_example_mesh("10x10_plane_hires")

# Specify coil plane geometry
center_offset = np.array([0, 0, 0])
standoff = np.array([0, 0.5, 0])

# Create coil plane pairs
coil_plus = trimesh.Trimesh(
    planemesh.vertices + center_offset + standoff, planemesh.faces, process=False
)
coil_minus = trimesh.Trimesh(
    planemesh.vertices + center_offset - standoff, planemesh.faces, process=False
)
joined_planes = combine_meshes((coil_plus, coil_minus))
# Create mesh class object
coil = MeshConductor(mesh_obj=joined_planes, fix_normals=True, basis_name="inner")
```
Here I define my target point. One point placed in the center of the design space and with the magnetic field in the positive X direction:
```python
target_points = np.zeros((1, 3))
target_field = np.zeros_like(target_points)
target_field[:, 0] = 1
```

```python
target_abs_error = np.zeros_like(target_field)
target_abs_error[:, 1] += 0.1
target_abs_error[:, 0::2] += 0.1
```

```python
#%%
# Plot target points and mesh
coil.plot_mesh(opacity=0.1)
mlab.quiver3d(*target_points.T, *target_field.T)

#%%
# Compute coupling matrix that is used to compute the generated magnetic field, create field specification
target_spec = {
    "coupling": coil.B_coupling(target_points),
    "abs_error": target_abs_error,
    "target": target_field,
}

#%%
# Run QP solver, plot result
import mosek
coil.s, prob = optimize_streamfunctions(
    coil,
    [target_spec],
    objective="minimum_inductive_energy",
    solver="MOSEK",
    solver_opts={"mosek_params": {mosek.iparam.num_threads: 8}},
)

coil.s.plot()
mlab.quiver3d(*target_points.T, *target_field.T)
```

Now the solution is found, so I create a plane of points in the XY plane, where I want to show the magnetic field: 
```python
x = y = np.linspace(-5, 5, 40)
X, Y = np.meshgrid(x, y, indexing="ij")

points = np.zeros((X.flatten().shape[0], 3))
points[:, 0] = X.flatten()
points[:, 1] = Y.flatten()
```

Here in the first method I calculate and plot the magnetic field using the method from the Biplanar coil design example:
```python
#%% METHOD 1
loops = coil.s.discretize(N_contours=6)
loops.plot_loops()
mlab.quiver3d(*target_points.T, *target_field.T, mode="arrow", scale_factor=0.3)

B = loops.magnetic_field(points)
mlab.quiver3d(*points.T, *B.T)
B_target = loops.magnetic_field(target_points)
mlab.quiver3d(*target_points.T, *B_target.T)
mlab.text3d(0, -4, 3, "METHOD 1: via loops.magnetic_field()", scale=0.3)
mlab.orientation_axes()
```
And here in the second method I calculate and plot the magnetic field using the method from the Magnetically shielded coil example:
```python
#%% METHOD 2
loops = scalar_contour(coil.mesh, coil.s.vert, N_contours=6)
plot_3d_current_loops(loops, tube_radius=0.01)
mlab.quiver3d(*target_points.T, *target_field.T, mode="arrow", scale_factor=0.3)

B = coil.B_coupling(points) @ coil.s
mlab.quiver3d(*points.T, *B.T)
B_target = coil.B_coupling(target_points) @ coil.s
mlab.quiver3d(*target_points.T, *B_target.T)
mlab.text3d(0, -4, 3, "METHOD 2: via coil.B_coupling(points) @ coil.s", scale=0.3)
mlab.orientation_axes()
```
Following METHOD 1 from above the calculated magnetic field in the XY plane is in the opposite direction compared to the target field. But the magnetic field in the XY plane makes sense with respect to the right hand rule on the current coils.
Following METHOD 2 from above, the calculated magnetic field in the XY plane is in the same direction as the target field. But the magnetic field does not follow the right hand rule on the current coils - every time I try, my thumb is almost dislocating.

If I use METHOD 1 from above, and do a `loops.reverse()` just after discritizing the stream function, then the magnetic field in the target point is directed in the positive X direction as expected and the right hand rule on the current coils feels much better.

Are the current in the resulting current loops having the correct direction?

**Comment by rzetter** [comment]
Thank you for noticing this bug, definitely is an issue in bfieldtools! On a first glance, it seems like scalar_contour() is to blame.

Method 1: coil.s.discretize() uses scalar_contour() to create a number of current loops from the stream function. The direction of these loops is incorrect (reversed). Thus, the field is in the incorrect direction but the right-hand rule is followed. 

Method 2: The plotted current loops are, as in method 1, derived from scalar_contour(). The computed magnetic field however is directly computed from the stream function, i.e. the continuous current density on the mesh planes. Thus the right-hand rule is not followed.

I'll try to reproduce the bug in other scenarios as well, will report back further.

@anttimc, is this something you would also have the time to look at?

**Comment by anttimc** [comment]
Yes, I could take a look at the code. If you can @rzetter create a minimal testcase to reproduce the issue, it would be a great help for debugging. If I understood correctly, the issue can be corrected with a sign flip in scalar_contour, but I have to check at which point.

**Comment by anttimc** [comment]
FYI @svejgaard @rzetter 

I'm working on the test case for scalar_contour now. I also took a look into the function itself. It's something I coded quite quickly and there are a couple of places that may cause the sign flip. First, I'll try to make this fix and commit it with the test case. After that let's see if we could refactor the code a bit. Although the comments attempt to explain what's happening, the code is a bit cryptic. I remember that at some point it was tested against mayavi/vtk contouring, but the direction of the contours was never tested.

**Comment by anttimc** [comment]
@svejgaard please check that your code works now. I've only checked that the commit fixes the new test case.

**Comment by svejgaard** [comment]
@rzetter @anttimc 
Thank you very much both of you.
Seems to work on my side. I tested using same approach from my initial post having a single target point in the center of the design space. I tried the 6 different scenarios having a unit magnetic field pointing in either of the X, Y, Z direction, both positively and negatively directed. All resulted in expected current direction and expected magnetic field.
I will write back here in case I see something unexpected with relation to this.
